### PR TITLE
Jobs and quest refactoring

### PR DESCRIPTION
**Pull request by ezsh**
1. Replace the QuestEngine class with Quest and QuestScene classes.
2. Refactor the common part of the Job and Quest classes into the Task
class.
3. Unify conditions checks and post effects processing between jobs and
quests.
4. Make quests work much like jobs, i.e. playing scenes (intro, middle,
finish).
5. Probably something else.

### CODE REVIEW

**Comment by ezsh** [comment]
@JJM152, my PRs conflict with each other already. Could you give an estimate when you can have a look and approve or reject at least some of them, please?

**Comment by ezsh** [comment]
Rebased. Although, don't quite remember how did I plan to continue.

**Comment by JJM152** [comment]
What to do about this? 

I hate tossing work, but as I mentioned in that other thread #130 - I'm thinking that I out smarted myself trying to come up with a format that I could throw a JSON editor on. 

Since you asked there what my reasoning is, I'll respond here - 

Originally I intended to use a simple JSON editor to build the quest and job content in order to speed up authoring, but as I did more and more repetitive tasks and kept expanding the job engine, I started to think that perhaps everything would have been better handled as just simple Twine passages with inline code and some support widgets and methods for some common tasks like checks, pushing flags and choosing rewards. 

There are a lot of virtues to working within the story format in terms of simplicity and flexibility of display of passages and content that are just lacking from the job engine implementation (and lets not talk about the shitty quest engine because I literally have not looked at it once since I slapped it in just to have something running).

Anyway, my thinking now is that I'm going to finish the game content within the limitations of the engine as it stands right now, primarily because I want to get this game finished and in a fully playable state so that I can move on to another project. I'm concerned that you'll deliver a shiny new implementation that simply isn't really needed, or even worse, I'll be left having to make a decision to refactor all of the existing json data to make it work correctly.

So, basically I'm going to leave this up to you - If you want to take on the work of jazzing this up as well as converting all the content and making sure it works, then I won't have any reason to say no at all, but if it seems like you have better things to do, then do them. The only caveat I have here is that I feel that the json schema for the jobs is already a shit fight, so I don't want it to get worse and more complicated. That's one of the reasons why I'm wishing I had just worked within the Twine story/macro format to start with. :cry:

**Comment by ezsh** [comment]
I see, thank you for the detailed explanation. Continuing to another project (also Twee/SugarCube (SC) game) which uses the direct approach, a simple and plain SC script for scenes, I can't tell that one is better. In the end that produces so many  SC variables it gets hard to maintain them. Your approach, from the other hand, your scenes-based approach is safe from all that mess. In general, among many Twine games I've seen this one is the most readable and well structured.

**Comment by ezsh** [comment]
Resurrected the PR. To me these changes make sense as they unify tests and rewards processign for quests and jobs.

**Comment by ezsh** [comment]
This PR is all about unifying quests and jobs handling, and now it achieves that with the following parts: checks and rewards, scene-based decomposition. Quests are converted to jobs with the fixed set of scenes (intro, middle, end) and are shown to the player as jobs. Due to the unification jobs gained the possibility to offer item choices as payment.

I did an extensive playthrough with these changes earlier, but due to many little changes to the jobs recently it very well can misbehave at points. But this should not, I believe, prevent reviewing.

File: src/000-SCRIPT_OBJ/JobEngine.js [hunk 1]
```diff
@@ -3,1298 +3,156 @@
...
+    }
 
     /**
      * @param {App.Entity.Player} Player
      * @param {string} Flag
      * @param {*} Value
      */
-    this.SetJobFlag = function(Player, Flag, Value)
-    {
+    SetJobFlagn(Player, Flag, Value) {
```
Comment (JJM152) [inline src/000-SCRIPT_OBJ/JobEngine.js]: Typo??

**Comment by JJM152** [comment]
Wow, I'm scared of this. I'll merge it tomorrow when I'm more awake and can devote some time to checking it out.

As an aside, what do you think of this: 
https://code.visualstudio.com/docs/languages/json#_mapping-to-a-schema-in-the-workspace

I've had a long open ticket #119 but I'm wondering if both a) the schema could be complex enough to handle the stupidity that is how the data entries are set up, or if vs code is the right tool to do this in and not something else. Thoughts?

File: src/000-SCRIPT_OBJ/JobEngine.js (hunk 1 above)
Comment (ezsh) [reply src/000-SCRIPT_OBJ/JobEngine.js]: Yes, thanks.

**Comment by ezsh** [comment]
> As an aside, what do you think of this:

I like the idea. Never tried schemes for JSON, but used them for XML with VS and it was very convenient. Browsing some time ago the AWS site that hosts VS Code default schemas I noticed the VS Code uses some really large scheme files (and apparently it works well).

**Comment by ezsh** [comment]
Doing a non-cheating playthrough with this branch again, I notice the cookie almost never rewards with the loot boxes (its lust is almost never high enough), while a few months ago the boxes were the drive force for the initial transformations. Checking with the jobs definitions I conclude the lust should be low, as it is with this branch. Am I wrong or there was a recent change to that (which I can't find in the git logs)?

**Comment by ezsh** [comment]
Was my fault. Testing a fix.

**Comment by ezsh** [comment]
> Testing a fix.

Seems to be OK now.

**Comment by JJM152** [comment]
I'll merge this now... cross fingers.

**Comment by ezsh** [comment]
God save the queen

**Comment by JJM152** [comment]
Can't complete Kiplers first quest (first real quest I guess in the game) - Tools of the Trade Pt 1

**Comment by ezsh** [comment]
Sorry. I'll take a look in a quarter of an hour.

**Comment by ezsh** [comment]
For the record, removing the condition makes it possible to complete the quest.
```diff
diff --git a/src/000-SCRIPT_OBJ/Tasks.js b/src/000-SCRIPT_OBJ/Tasks.js
index a8980c3..3e55114 100644
--- a/src/000-SCRIPT_OBJ/Tasks.js
+++ b/src/000-SCRIPT_OBJ/Tasks.js
@@ -143,7 +143,7 @@ App.Task = class Task {
                     }
                     break;
                 case "NPC_STAT":
-                    if (typeof NPC === 'undefined' && typeof Option !== 'undefined') NPC = Player.GetNPC(Option);
+                    if (typeof Option !== 'undefined') NPC = Player.GetNPC(Option);
                     if (typeof NPC !== 'undefined') {
                         if (Task._Cmp( NPC.GetStat(Name), Value, Condition ) == false) {
                             StatusFlag = false;
```

**Comment by ezsh** [comment]
I think I also broke these two and alike:
```json
{ "TAG" : "B", "TYPE" : "SKILL", "NAME" : "Cooking", "DIFFICULTY" : 30, "REWARD" : "MONEY", "R_NAME" : "MONEY", "VALUE" : 10, "OPT" : 0 },
{ "TAG" : "C", "TYPE" : "SKILL", "NAME" : "AssFucking", "DIFFICULTY" : 30, "REWARD" : "MONEY","R_NAME" : "MONEY","VALUE" : 10, "OPT" : 0 }],
```
Do they give rewards in addition to setting tag values?

**Comment by ezsh** [comment]
See #191.

**Comment by JJM152** [comment]
yawn just woke up, have to go to work, I'll check this out around lunch time :)
